// arui-feather/theme-provider/theme-provider.js
'use strict';

exports.__esModule = true;

var _class, _temp; /* This Source Code Form is subject to the terms of the Mozilla Public
                    * License, v. 2.0. If a copy of the MPL was not distributed with this
                    * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

/**
 * Компонент задающий тему для своих дочерних компонентов.
 * Важно! Может содержать в себе строго один дочерний компонент.
 *
 * @example
 * ```javascript
 * import ThemeProvider from 'arui-feather/theme-provider';
 * import Page from 'arui-feather/page';
 * import Heading from 'arui-feather/heading';
 *
 * <ThemeProvider theme="alfa-on-color">
 *    <Page>
 *       <Heading>Заголовок страницы</Heading>
 *       <div style={{ background: "white" }}>
 *           <ThemeProvider theme="alfa-on-white">
 *               Врезка белого цвета на странице...
 *           </ThemeProvider>
 *       </div>
 *    </Page>
 * </ThemeProvider>
 * ```
 */
var ThemeProvider = (_temp = _class = function (_React$Component) {
    _inherits(ThemeProvider, _React$Component);

    function ThemeProvider() {
        _classCallCheck(this, ThemeProvider);

        return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
    }

    ThemeProvider.prototype.getChildContext = function getChildContext() {
        return {
            theme: this.props.theme
        };
    };

    ThemeProvider.prototype.render = function render() {
        return _react2.default.Children.only(this.props.children);
    };

    return ThemeProvider;
}(_react2.default.Component), _class.propTypes = {
    /** Дочерний элемент `ThemeProvider` */
    children: _propTypes2.default.node,
    /** Дополнительный класс */
    className: _propTypes2.default.string,
    /** Тема компонента */
    theme: _propTypes2.default.oneOf(['alfa-on-color', 'alfa-on-white'])
}, _class.contextTypes = {
    theme: _propTypes2.default.string
}, _class.childContextTypes = {
    theme: _propTypes2.default.string
}, _temp);
exports.default = ThemeProvider;
