// arui-feather/progress-bar/progress-bar.js
'use strict';

exports.__esModule = true;

var _dec, _dec2, _class, _class2, _temp;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _cn = require('../cn');

var _cn2 = _interopRequireDefault(_cn);

var _performance = require('../performance');

var _performance2 = _interopRequireDefault(_performance);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defaults(obj, defaults) { var keys = Object.getOwnPropertyNames(defaults); for (var i = 0; i < keys.length; i++) { var key = keys[i]; var value = Object.getOwnPropertyDescriptor(defaults, key); if (value && value.configurable && obj[key] === undefined) { Object.defineProperty(obj, key, value); } } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : _defaults(subClass, superClass); }

/**
 * Компонент прогресс-бара.
 */
var ProgressBar = (_dec = (0, _cn2.default)('progress-bar'), _dec2 = (0, _performance2.default)(), _dec(_class = _dec2(_class = (_temp = _class2 = function (_React$Component) {
    _inherits(ProgressBar, _React$Component);

    function ProgressBar() {
        _classCallCheck(this, ProgressBar);

        return _possibleConstructorReturn(this, _React$Component.apply(this, arguments));
    }

    ProgressBar.prototype.render = function render(cn) {
        var styles = { width: this.props.percent + '%' };

        return _react2.default.createElement(
            'div',
            { className: cn({ size: this.props.size }) },
            _react2.default.createElement('div', {
                style: styles,
                className: cn('current-value')
            })
        );
    };

    return ProgressBar;
}(_react2.default.Component), _class2.propTypes = {
    /** Прогресс в процентах */
    percent: _propTypes2.default.number,
    /** Размер компонента */
    size: _propTypes2.default.oneOf(['m']),
    /** Дополнительный класс */
    className: _propTypes2.default.string
}, _class2.defaultProps = {
    percent: 0,
    size: 'm'
}, _temp)) || _class) || _class);
exports.default = ProgressBar;
